src/gtest/test_checktransaction.cpp

```diff
@@ -391,3 +391,346 @@ TEST(checktransaction_tests, non_canonical_ed25519_signature) {
+    ContextualCheckTransaction(tx, state, 1, 100);
+}
+
+// This tests a transaction without the fOverwintered flag set, against the Overwinter consensus rule set.
+TEST(checktransaction_tests, OverwinterFlagNotSet) {
+    SelectParams(CBaseChainParams::REGTEST);
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, Consensus::NetworkUpgrade::ALWAYS_ACTIVE);
+
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.fOverwintered = false;
+    mtx.nVersion = 3;
+    mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
+    mtx.nExpiryHeight = 0;
+
+    CTransaction tx(mtx);
+    MockCValidationState state;
+    EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "tx-overwinter-flag-not-set", false)).Times(1);
+    ContextualCheckTransaction(tx, state, 1, 100);
+
+    // Revert to default
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
+}
+
+
+// Contextual test an Overwintered transaction with a bad version number 
+TEST(checktransaction_tests, OverwinteredVersionTooLow) {
+    SelectParams(CBaseChainParams::REGTEST);
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, Consensus::NetworkUpgrade::ALWAYS_ACTIVE);
+
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.fOverwintered = true;
+    mtx.nVersion = 3;
+    mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
+    mtx.nExpiryHeight = 0;
+
+    {
+        CTransaction tx(mtx);
+
+        // Evil
+        *const_cast<int32_t*>(&tx.nVersion) = CTransaction::OVERWINTER_MIN_CURRENT_VERSION - 1;
+
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "tx-overwinter-version-too-low", false)).Times(1);
+        ContextualCheckTransaction(tx, state, 1, 100);
+    }
+
+    {
+        CTransaction tx(mtx);
+
+        // Evil
+        *const_cast<int32_t*>(&tx.nVersion) = CTransaction::OVERWINTER_MAX_CURRENT_VERSION + 1;
+
+        MockCValidationState state;
+        EXPECT_CALL(state, DoS(100, false, REJECT_INVALID, "tx-overwinter-version-too-high", false)).Times(1);
+        ContextualCheckTransaction(tx, state, 1, 100);
+    }
+
+    // Revert to default
+    UpdateNetworkUpgradeParameters(Consensus::UPGRADE_OVERWINTER, Consensus::NetworkUpgrade::NO_ACTIVATION_HEIGHT);
+}
+
+
+// Overwinter (NU0) does not allow soft fork to version 4 Overwintered tx.
+TEST(checktransaction_tests, OverwinterInvalidSoftForkVersion) {
+    CMutableTransaction mtx = GetValidTransaction();
+    mtx.fOverwintered = true;
+    mtx.nVersion = 4; // This is not allowed
+    mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
+    mtx.nExpiryHeight = 0;
+
+    CDataStream ss(SER_DISK, PROTOCOL_VERSION);
+    try {
+        ss << mtx;
+        FAIL() << "Expected std::ios_base::failure";
+    }
+    catch(std::ios_base::failure & err) {
+        EXPECT_EQ(err.what(), std::string("Unknown transaction format: iostream error"));
```

Comment:
`EXPECT_THAT(err.what(), HasSubstr(std::string("Unknown transaction format")))` (see the [CheatSheet](https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md) for details).

Reply:
Bookmarked!